packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts

```diff
@@ -0,0 +1,184 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { analyticsToAnalyticClusters } from '@tupaia/data-broker';
+import { getUniqueEntries } from '@tupaia/utils';
+import { AnalyticsRepository } from '../../AnalyticsRepository';
+import { getExpressionParserInstance } from '../../getExpressionParserInstance';
+import { Aggregation, Analytic, AnalyticCluster, FetchOptions, Indicator } from '../../types';
+import { Builder } from '../Builder';
+import { createBuilder } from '../createBuilder';
+import { getElementCodesForBuilders } from '../utils';
+import {
+  ArithmeticConfig,
+  DefaultValue,
+  configValidators,
+  getAggregationsByCode,
+  isParameterCode,
+} from './config';
+
+/**
+ * Config used by the builder. It is essential a fully expanded, verbose version
+ * of the indicator config passed in by the user (`ArithmeticConfig`)
+ */
+type BuilderConfig = {
+  readonly formula: string;
+  readonly aggregation: Record<string, Aggregation[]>;
+  readonly parameters: Indicator[];
+  readonly defaultValues: Record<string, DefaultValue>;
+};
+
+const indicatorToBuilderConfig = (indicatorConfig: ArithmeticConfig): BuilderConfig => {
+  const { defaultValues = {}, parameters = [], ...otherFields } = indicatorConfig;
+
+  return {
+    ...otherFields,
+    defaultValues,
+    parameters,
+    aggregation: getAggregationsByCode(indicatorConfig),
+  };
+};
+
+export class ArithmeticBuilder extends Builder {
+  private configCache: BuilderConfig | null = null;
+
+  get config() {
+    if (!this.configCache) {
+      const config = this.validateConfig<ArithmeticConfig>(configValidators);
+      this.configCache = indicatorToBuilderConfig(config);
+    }
+    return this.configCache;
+  }
+
+  getElementCodes = (): string[] => {
+    const codesInFormula = this.getElementCodesInFormula();
+    const codesInParameters = this.getElementCodesInParameters();
+    return getUniqueEntries(codesInParameters.concat(codesInFormula));
+  };
+
+  private getElementCodesInFormula = () =>
+    Object.keys(this.config.aggregation).filter(
+      variable => !isParameterCode(this.config.parameters, variable),
+    );
+
+  private getElementCodesInParameters = () => {
+    const parameterBuilders = this.config.parameters.map(p => createBuilder(p));
+    return getElementCodesForBuilders(parameterBuilders);
+  };
+
+  getAggregations = (): Aggregation[] => {
+    const formulaAggregations = Object.values(this.config.aggregation).flat();
+    const parameterAggregations = this.config.parameters
+      .map(p => createBuilder(p).getAggregations())
```

Comment:
Great point @edmofro ! I also thought I could introduce a `parameterBuilderCache` (similar to the `configCache` in this class). But I have the impression that this operation is fairly cheap anyway.

I may attempt to do so and get an actual measurement. I suspect the performance diff will be negligible (although you never know!), but it may even make the code a bit simpler!

Reply:
Eventually I did use a parameter builder cache, since it seems to simplify code!